util/resolver/authorizer.go

```diff
@@ -166,6 +166,11 @@ func (a *dockerAuthorizer) AddResponses(ctx context.Context, responses []*http.R
 				}
 			}
 
+			// prevent cached expired (or otherwise newly invalid) tokens from being reused
+			if handler != nil && last.StatusCode == http.StatusUnauthorized {
+				handler.deleteLastScopedTokenResult(ctx)
+			}
```

Comment:
Another option might be to surface the last status code (or some other error-signifying value) into the handler. We could then check it along with the expiration check to see if we should fetch a new token ([somewhere around here, maybe](https://github.com/moby/buildkit/blob/df49b648c8bfef657ba99543390516a3f518d22a/util/resolver/authorizer.go#L298-L302))?